src/gui/editorwidgets/qgscheckboxwidgetwrapper.cpp

```diff
@@ -64,12 +76,16 @@ void QgsCheckboxWidgetWrapper::initWidget( QWidget *editor )
   mGroupBox = qobject_cast<QGroupBox *>( editor );
 
   if ( mCheckBox )
-    connect( mCheckBox, &QAbstractButton::toggled, this, [ = ]( bool state )
+    connect( mCheckBox, &QCheckBox::stateChanged, this, [ = ]( int state )
   {
+    if ( !mIndeterminateStateEnabled && mCheckBox->checkState() != Qt::PartiallyChecked )
+    {
+      mCheckBox->setTristate( false );
```

Comment:
this prevents "undoing" a null -> not null change, right? Do you think we should just leave it as tristate so a user can undo this?

Reply:
(if so, then the valueChanged signal logic below needs to handle null values too)

Reply:
@nyalldawson , I'm -1 to introduce a _settable_ null state via tristate. For one, it'll interfere with multi-editing, where the PartiallyChecked state is used to say "don't change anything". If we want a settable null value in the future, we can implement a separate set-to-null button. But that's beyond my scope of interest with this PR (i.e. fixing the constraint issue).